From 26d1373a403790a57d1647ea713d6c4d35a40a44 Mon Sep 17 00:00:00 2001
From: Hu Kejun <william.hu@rock-chips.com>
Date: Mon, 20 Aug 2018 18:17:08 +0800
Subject: [PATCH] media: rockchip: isp1: fix sp will stop to output after mp
 start in isp12

fix when sp and mp are configured to output the same size, sp will stop
to output after mp start in isp12.

Change-Id: Iec07aa8116ead608999f02dcb357f2dd2997d4a9
Signed-off-by: Hu Kejun <william.hu@rock-chips.com>
---
 drivers/media/platform/rockchip/isp1/regs.h   | 30 +++++++++++++++++++
 drivers/media/platform/rockchip/isp1/rkisp1.c |  8 +++++
 2 files changed, 38 insertions(+)

diff --git a/drivers/media/platform/rockchip/isp1/regs.h b/drivers/media/platform/rockchip/isp1/regs.h
index a18a7b52e1bb..1b658c615f0f 100644
--- a/drivers/media/platform/rockchip/isp1/regs.h
+++ b/drivers/media/platform/rockchip/isp1/regs.h
@@ -240,6 +240,35 @@
 
 /* CCL */
 #define CIF_CCL_CIF_CLK_DIS			BIT(2)
+/* VI_ISP_CLK_CTRL */
+#define CIF_CLK_CTRL_ISP_RAW			BIT(0)
+#define CIF_CLK_CTRL_ISP_RGB			BIT(1)
+#define CIF_CLK_CTRL_ISP_YUV			BIT(2)
+#define CIF_CLK_CTRL_ISP_3A			BIT(3)
+#define CIF_CLK_CTRL_MIPI_RAW			BIT(4)
+#define CIF_CLK_CTRL_ISP_IE			BIT(5)
+#define CIF_CLK_CTRL_RSZ_RAM			BIT(6)
+#define CIF_CLK_CTRL_JPEG_RAM			BIT(7)
+#define CIF_CLK_CTRL_ACLK_ISP			BIT(8)
+#define CIF_CLK_CTRL_MI_IDC			BIT(9)
+#define CIF_CLK_CTRL_MI_MP			BIT(10)
+#define CIF_CLK_CTRL_MI_JPEG			BIT(11)
+#define CIF_CLK_CTRL_MI_DP			BIT(12)
+#define CIF_CLK_CTRL_MI_Y12			BIT(13)
+#define CIF_CLK_CTRL_MI_SP			BIT(14)
+#define CIF_CLK_CTRL_MI_RAW0			BIT(15)
+#define CIF_CLK_CTRL_MI_RAW1			BIT(16)
+#define CIF_CLK_CTRL_MI_READ			BIT(17)
+#define CIF_CLK_CTRL_MI_RAWRD			BIT(18)
+#define CIF_CLK_CTRL_CP				BIT(19)
+#define CIF_CLK_CTRL_IE				BIT(20)
+#define CIF_CLK_CTRL_SI				BIT(21)
+#define CIF_CLK_CTRL_RSZM			BIT(22)
+#define CIF_CLK_CTRL_DPMUX			BIT(23)
+#define CIF_CLK_CTRL_JPEG			BIT(24)
+#define CIF_CLK_CTRL_RSZS			BIT(25)
+#define CIF_CLK_CTRL_MIPI			BIT(26)
+#define CIF_CLK_CTRL_MARVINMI			BIT(27)
 /* ICCL */
 #define CIF_ICCL_ISP_CLK			BIT(0)
 #define CIF_ICCL_CP_CLK				BIT(1)
@@ -693,6 +722,7 @@
 #define CIF_CTRL_BASE			0x00000000
 #define CIF_CCL				(CIF_CTRL_BASE + 0x00000000)
 #define CIF_VI_ID			(CIF_CTRL_BASE + 0x00000008)
+#define CIF_VI_ISP_CLK_CTRL_V12		(CIF_CTRL_BASE + 0x0000000C)
 #define CIF_ICCL			(CIF_CTRL_BASE + 0x00000010)
 #define CIF_IRCL			(CIF_CTRL_BASE + 0x00000014)
 #define CIF_VI_DPCL			(CIF_CTRL_BASE + 0x00000018)
diff --git a/drivers/media/platform/rockchip/isp1/rkisp1.c b/drivers/media/platform/rockchip/isp1/rkisp1.c
index 54f2e04f51cf..ea467b92c07d 100644
--- a/drivers/media/platform/rockchip/isp1/rkisp1.c
+++ b/drivers/media/platform/rockchip/isp1/rkisp1.c
@@ -474,6 +474,14 @@ static void rkisp1_config_clk(struct rkisp1_device *dev)
 		CIF_ICCL_IE_CLK | CIF_ICCL_MIPI_CLK | CIF_ICCL_DCROP_CLK;
 
 	writel(val, dev->base_addr + CIF_ICCL);
+
+	if (dev->isp_ver == ISP_V12) {
+		val = CIF_CLK_CTRL_MI_Y12 | CIF_CLK_CTRL_MI_SP |
+		      CIF_CLK_CTRL_MI_RAW0 | CIF_CLK_CTRL_MI_RAW1 |
+		      CIF_CLK_CTRL_MI_READ | CIF_CLK_CTRL_MI_RAWRD |
+		      CIF_CLK_CTRL_CP | CIF_CLK_CTRL_IE;
+		writel(val, dev->base_addr + CIF_VI_ISP_CLK_CTRL_V12);
+	}
 }
 
 /***************************** isp sub-devs *******************************/
-- 
2.35.3

